### ISSUE TITLE
bcf_hdr_read() and vcf_hdr_read() download index as side-effect

### ISSUE THREAD

**Issue by jeromekelleher** [samtools/htslib#380]
Calling `bcf_header_read` results in an index file being downloaded to the local directory, which seems like an undesirable side effect of a function that (I assume) reads the header of a VCF/BCF file. This behaviour causes trouble in the GA4GH reference server --- see  https://github.com/pysam-developers/pysam/issues/286 and https://github.com/ga4gh/server/issues/1210.

This code snippet is derived from the pysam source code and should make it easy to reproduce the problem.

``` C
#include <config.h>

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <getopt.h>
#include <unistd.h>

#include "htslib/hfile.h"
#include "htslib/hts.h"
#include "htslib/tbx.h"
#include "htslib/vcf.h"


int main(int argc, char **argv)
{

    int status = 0;
    const char *file = "http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/integrated_sv_map/ALL.autosomes.pindel.20130502.complexindex.low_coverage.genotypes.vcf.gz";
    const char *index_file = "/tmp/ALL.autosomes.pindel.20130502.complexindex.low_coverage.genotypes.vcf.gz.tbi";
    htsFile *hp = NULL;
    tbx_t *index = NULL;
    bcf_hdr_t *header = NULL;

    hp = hts_open(file, "r");
    if (hp == NULL) {
        fprintf(stderr, "can't open \"%s\": %s\n", file, strerror(errno));
        status = EXIT_FAILURE;
        goto out;
    }

    header = bcf_hdr_read(hp);
    if (header == NULL) {
        fprintf(stderr, "error reading header \"%s\": %s\n", file, strerror(errno));
        status = EXIT_FAILURE;
        goto out;
    }

    index = tbx_index_load2(file, index_file);
    if (index == NULL) {
        fprintf(stderr, "can't open \"%s\": %s\n", index_file, strerror(errno));
        status = EXIT_FAILURE;
        goto out;
    }
out:
    return status;
}
```

Is this the intended behaviour of `bcf_hdr_read`? If so, how can we work around this issue? Side effects like this are really horrible for a server --- see the discussion in #375.

**Comment by jeromekelleher** [comment]
Any thoughts on this @jmarshall/@jkbonfield? It would be good to have some idea of how difficult this will be to fix, so we can decide on whether we should figure out workarounds or just disable our remote VCF support.

**Comment by jkbonfield** [comment]
This isn't an area of the code I'm currently working on, but maybe one for John or Rob?

**Comment by jmarshall** [comment]
For VCF files, these functions [augment the headers returned with any extra contigs listed in the index](https://github.com/samtools/htslib/blob/1.3.1/vcf.c#L1338-L1359).

This is somewhat conceptually dubious in general, though I can see the merge simplifications it enables when dealing with several samples with variants on different subsets of the same contigs.

I think we all agree with your “creating local files sucks for daemon processes” comments on #375, so this is something that should be fixed.

Perhaps in the short term the functionality could be made conditional on `hisremote()`.  Really what we want is fully separated load/save APIs so we could say “read in the index but don't save it anywhere” to do this header augmentation.  Or even a way to read in just the contig headers from the index.

In short, this should be fixed but I don't think it will be a high priority.

**Comment by daviesrob** [comment]
I think the bigger question here is should we cache the index files at all?  The way it's done at the moment is not good.  There are race conditions and the possibility of using the wrong cached file as its name happens to match the one being downloaded.

Most of the cached files seem to be small, for example the biggest .fai file I've found so far is about 34 Mb (for a Tasmanian devil assembly). The corresponding .gzi file is less than 1Mb.  Most .fai files are much smaller, in our reference collection more than half are less than 100 bytes.

.tbi are also small.  For example, I have a 5Gb vcf.gz file with a 223K .tbi file.  For all of these indexes, htslib reads the entire file and doesn't try to do any random access.

I now have a version of htslib that can read remote .fai and .gzi files directly without having to save them first.  Given that it's fairly easy to do (just replace stdio with hfile), maybe we should stop trying to cache files and encourage alternative solutions (e.g. web proxies) instead?

**Comment by jeromekelleher** [comment]
Thanks for the explanation @jmarshall, this is very useful. I think the best thing for us to do then is just disable remote VCF support --- it's nowhere near as important as remote BAM support, so it can wait until the issue is fixed.

@daviesrob absolutely agree that index files shouldn't be cached, and should only be opened when it's explicitly requested.

**Comment by valeriuo** [comment]
If this is still an issue, we can now drop the `HTS_IDX_SAVE_REMOTE` flag from the `tbx_index_load3` [call](https://github.com/samtools/htslib/blob/6038f97e901cc2be832cc7749200faf375b287e4/vcf.c#L1919) or make it optional.
